libraries/chain/controller.cpp

```diff
@@ -3999,15 +3986,95 @@ struct controller_impl {
+                  "Transition block #${b} has a strong QC claim",
+                  ("b", block_num) );
+      EOS_ASSERT( !f_ext.new_proposer_policy_diff,
+                  block_validate_exception,
+                  "Transition block #${b} has new_proposer_policy_diff",
+                  ("b", block_num) );
+
+      if (auto it = prev.header_exts.find(finality_extension::extension_id()); it != prev.header_exts.end()) {
+         const auto& prev_finality_ext = std::get<finality_extension>(it->second);
+         EOS_ASSERT( f_ext.qc_claim.block_num == prev_finality_ext.qc_claim.block_num,
+                     block_validate_exception,
+                     "Transition block #${b} QC claim block_num not equal to previous QC claim block_num",
+                     ("b", block_num) );
+      } else {
+         // Savanna Genesis Block
+         EOS_ASSERT( f_ext.qc_claim.block_num == block_num,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} QC claim block_num not equal to current block_num",
+                     ("b", block_num) );
+         EOS_ASSERT( f_ext.new_finalizer_policy_diff,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} finality block header extension misses new_finalizer_policy_diff",
+                     ("b", block_num) );
+
+         finalizer_policy no_policy;
+         // apply_diff will FC_ASSERT if new_finalizer_policy_diff is malformated
+         finalizer_policy genesis_policy = no_policy.apply_diff(*f_ext.new_finalizer_policy_diff);
+         EOS_ASSERT( genesis_policy.generation == 1,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} finalizer policy generation (${g}) not 1",
+                     ("b", block_num)("g", genesis_policy.generation) );
+      }
+   }
+
    // thread safe, expected to be called from thread other than the main thread
    template<typename ForkDB, typename BS>
    block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
       constexpr bool savanna_mode = std::is_same_v<typename std::decay_t<BS>, block_state>;
+
+      // Verify claim made by finality_extension in block header extension and
+      // quorum_certificate_extension in block extension are valid.
+      // This is the only place the evaluation is done.
+      // Note: the purpose of running verify_qc_claim on Legacy blocks too is to
+      // safe guard Legacy blocks.
       if constexpr (savanna_mode) {
-         // Verify claim made by finality_extension in block header extension and
-         // quorum_certificate_extension in block extension are valid.
-         // This is the only place the evaluation is done.
          verify_qc_claim(id, b, prev);
+      } else {
+         verify_legacy_block_exts(b, prev);
       }
 
       auto trx_mroot = calculate_trx_merkle( b->transactions, savanna_mode );
```

Comment:
I don't think we actually care about `savanna_mode` here.

What we really care about is if `b` is a Proper Savanna block or not. We can determine that using `b->is_proper_svnn_block()`.

And if we add the asserts I mentioned above earlier, we would have confidence that this is doing the same thing as passing in `savanna_mode`.

Reply:
`is_proper_svnn_block()` is better. This was before we had that method.

Reply:
I still go back to having two functions with the same name:

```
// called for all non-proper blocks received
verify_block_extensions(const signed_block_ptr& b, const block_state_legacy& prev);

// called for all proper blocks received
verify_block_extensions(const signed_block_ptr& b, const block_state& prev);
```

and just call:

```
   block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
      verify_block_extensions(b, prev);
...
```

Reply:
We would still need to special case the other `if constexpr` below that integrates a QC and considers voting. These only make sense with a `block_state`.

Not to mention that the future optimizations that delay the actual verification of the attached QC to the chain threads.

Reply:
However, we could at least rename `savanna_mode` to `is_proper_savanna_block` and add the assert at the top:

```cpp
assert(is_proper_savanna_block == b->is_proper_svnn_block());
```